### PR TITLE
routing: add OLSRv2

### PR DESCRIPTION
**Pull request by OlegHahm**
In consultation with @benpicco I take over PR #765. Original description:

--
This adds an implementation of the OLSRv2 routing protocol (without metrics), based on oonf_api.
The interface offered to applications currently consists of only a olsr_init function which will set up and start a sender and receiver thread for RFC5444 based (UDP on port 269) HELLO and Topology Control messages.
That means that every HELLO_REFRESH_INTERVAL (2s + random jitter) a non-forwarded HELLO message is send for neighborhood discovery. Every TC_REFRESH_INTERVAL (5s + random jitter) a topology control message is broadcast (only by selected flooding MPRs) through the network.

This happens transparent to the application, IP packets are now routed to known hosts within the MANET once a path has been established.

print_topology_set will output the forwarding table with all known mesh nodes.

Does work well on native with #680, but still crashes on the testbed after some time.
Each known MANET router requires 72 byte of memory + 12 bytes for each alternative route to the node.

Depends on #630

### CODE REVIEW

**Comment by cgundogan** [comment]
are there plans to remove the RFC5444 code from this PR and use the NHDP implementation from @fnack instead?

**Comment by OlegHahm** [comment]
Might make sense indeed. Will try to check with @fnack.

**Comment by OlegHahm** [comment]
Rebased.

File: Makefile.dep [hunk 1]
```diff
@@ -2,6 +2,16 @@ ifneq (,$(filter libcoap,$(USEPKG)))
 	USEMODULE += pnet
 endif
 
+ifneq (,$(filter olsr2,$(USEMODULE)))
+	USEMODULE += vtimer
+	USEMODULE += socket_base
```
Comment (cgundogan) [inline Makefile.dep]: remove this line. udp has it as a dependency in Makefile.dep

File: sys/slist/Makefile [hunk 2]
```diff
@@ -0,0 +1 @@
+include $(RIOTBASE)/Makefile.base
```
Comment (cgundogan) [inline sys/slist/Makefile]: moving the slist implementation into its own PR and then rebase the olsr branch onto the slist branch would reduce the diff a little bit..

File: core/include/kernel_macros.h [hunk 3]
```diff
@@ -16,6 +16,9 @@
  * @author      René Kijewski
  */
 
+#ifndef KERNEL_MACROS_
```
Comment (cgundogan) [inline core/include/kernel_macros.h]: the diff of this file could be completely moved to its own PR. seems unrelated to OLSR to me.

File: pkg/oonf_api/0007-Use-RIOT-s-container_of-implementation.patch [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
++++ b/src-api/common/container_of.h
+@@ -59,10 +59,24 @@
+  * @return pointer to surrounding struct
+  */
+ #ifndef container_of
+-#define container_of(ptr, type, member) ({ \
+-    const typeof(((type *)0)->member ) *__tempptr = (ptr); \
+-    (type *)((char *)__tempptr - offsetof(type,member)); \
+-  })
++#if __STDC_VERSION__ >= 201112L
```
Comment (cgundogan) [inline pkg/oonf_api/0007-Use-RIOT-s-container_of-implementation.patch]: quick question: why is this the same as in *core/include/kernel_macros.h*. Is it possible to reduce the redundance anywhere?

File: sys/Makefile [hunk 5]
```diff
@@ -10,6 +10,21 @@ endif
 ifneq (,$(filter shell_commands,$(USEMODULE)))
     DIRS += shell/commands
 endif
+ifneq (,$(filter slist,$(USEMODULE)))
```
Comment (cgundogan) [inline sys/Makefile]: IMHO all following lines (excluding the olsr specific ones) are unnecessary in this file

File: sys/net/routing/olsr2/constants.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+/* in seconds */
+#define OLSR2_HELLO_REFRESH_INTERVAL_SECONDS	2
+#define OLSR2_TC_REFRESH_INTERVAL_SECONDS	5
+#define OLSR2_HOLD_TIME_SECONDS		(3 * OLSR2_TC_REFRESH_INTERVAL_SECONDS)
+
+#define OLSR2_TC_HOP_LIMIT	16
+
+#define OLSR2_HYST_SCALING	0.4
+#define OLSR2_HYST_LOW		0.3
+#define OLSR2_HYST_HIGH		0.8
```
Comment (cgundogan) [inline sys/net/routing/olsr2/constants.h]: -1 indentation

File: core/include/kernel_macros.h (hunk 3 above)
Comment (OlegHahm) [inline core/include/kernel_macros.h]: :+1:

File: sys/net/routing/olsr2/constants.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+/* in µs */
+#define SECOND			(1000 * 1000)
+#define OLSR2_MAX_JITTER_US	SECOND
```
Comment (cgundogan) [inline sys/net/routing/olsr2/constants.h]: s/SECOND/SEC_IN_USEC/g

File: sys/net/routing/olsr2/olsr_init.c [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    get_local_addr()->_prefix_len = 128;
+    ipv6_net_if_get_best_src_addr((ipv6_addr_t *) get_local_addr(), &sa_bcast.sin6_addr);
+
+    /* register olsr for routing */
+    ipv6_iface_set_routing_provider(get_next_hop);
+
+    DEBUG("This is node %s with IP %s", olsr2_local_name, netaddr_to_str_s(&nbuf[0], get_local_addr()));
+
+    /* enable sending */
+    int pid = thread_create(sender_thread_stack, sizeof sender_thread_stack, PRIORITY_MAIN - 1, CREATE_STACKTEST, olsr_sender_thread, NULL, "olsr_snd");
```
Comment (cgundogan) [inline sys/net/routing/olsr2/olsr_init.c]: Is it possible to combine both threads (l202[receiving] + l215[sending]) into only 1 thread?

File: sys/net/routing/olsr2/node.c [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+    return netaddr_cmp(a->addr, b->addr);
+}
+
+void node_init(void)
+{
+    local_addr._refs = 1;
+    avl_init(get_olsr_head(), _addr_cmp, false);
+}
+
+struct netaddr *get_local_addr(void)
```
Comment (cgundogan) [inline sys/net/routing/olsr2/node.c]: might make sense to inline this function or keep a global pointer already casted to &local_addr. This function gets called quite frequently.

File: sys/net/routing/olsr2/olsr_init.c (hunk 8 above)
Comment (OlegHahm) [inline sys/net/routing/olsr2/olsr_init.c]: Maybe, but I won't touch this in this PR.

**Comment by emmanuelsearch** [comment]
Is this still planned for 2015.12 ?

**Comment by OlegHahm** [comment]
Definitely not - most likely a mishap while retagging some PRs and issues.

**Comment by OlegHahm** [comment]
I have to admit: I was not very successful adopting this PR...
